### PR TITLE
[GraphBolt] add MinibatchSampler which supports ItemSet

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

MinibatchSampler for GraphBolt MVP. only support `ItemSet` for now. `DictItemSet` is to be supported.

current implementation is highly based on `torchdata.datapipes`. Both `ItemSet` and `MinibatchSampler` are required to be **Iterable** only, so it may not be as efficient as `dgl.dataloading.DataLoader` which operates on `torch.Tensor` which supports slicing.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    2. Node pairs.
+    >>> item_set = gb.ItemSet((torch.arange(0, 10), torch.arange(10, 20)))
+    >>> minibatch_sampler = gb.MinibatchSampler(
+    ...     item_set, batch_size=4, shuffle=True, drop_last=False
+    ... )
+    >>> list(minibatch_sampler)
+    [[tensor([9, 8, 3, 1]), tensor([19, 18, 13, 11])], [tensor([2, 5, 7, 4]),
+    tensor([12, 15, 17, 14])], [tensor([0, 6]), tensor([10, 16])]
+
+    3. Node pairs and labels.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:68]: TODO?

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    4. Head, tail and negative tails
+
+    5. DGLGraphs.
+
+    6. Append other datapipes.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:74]: This title seems not matching with the code below.
The code just added an enumerate(), do we have a better example or maybe just rename it with work with enumerate()?

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    counterparts.
+
+    Note: This class `MinibatchSampler` is not decorated with
+    `torchdata.datapipes.functional_datapipe` on purpose. This indicates it
+    does not support function-like call. But any iterable datapipes from
+    `torchdata` can be further appended.
+
+    Parameters
+    ----------
+    item_set : ItemSet or DictItemSet
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:38]: add an example for DictItemSet?

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        super().__init__()
+        self._item_set = item_set
+        self._batch_size = batch_size
+        self._drop_last = drop_last
+        self._shuffle = shuffle
+
+    def __iter__(self):
+        data_pipe = IterableWrapper(self._item_set)
+        if self._shuffle:
+            data_pipe = data_pipe.shuffle()
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:98]: Does the shuffle work if the iterable is a stream?

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        data_pipe = data_pipe.batch(
+            batch_size=self._batch_size,
+            drop_last=self._drop_last,
+        )
+        data_pipe = data_pipe.collate(collate_fn=_colloate)
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:103]: move the function code to L102 for better readability?

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 2 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:74]: let me choose another datapipe for illustration

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 3 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:38]: `DictItemSet` is not supported yet as I'd like to make this PR small enough. I'd like to add support for `DictItemSet` in another PR.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 4 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:98]: Yes. I reviewed the implementation in `torchdata`: https://github.com/pytorch/pytorch/blob/6ac3352a37ead6bae0de9fe4c45b618f8e82dc78/torch/utils/data/datapipes/iter/combinatorics.py#L122-L135

File: python/dgl/graphbolt/minibatch_sampler.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+from torch.utils.data import default_collate
+from torchdata.datapipes.iter import IterableWrapper, IterDataPipe
+from .itemset import *
+from ..batch import batch as dgl_batch
+from ..heterograph import DGLGraph
+
+__all__ = ["MinibatchSampler"]
+
+
+def _colloate(batch):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch_sampler.py:14]: typo:
coalesce? collate?
```suggestion
def _collate(batch):
```

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 3 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/minibatch_sampler.py:38]: Remove comment related to DictItemSet in this PR and add in the other one to avoid confusion.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 6 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:14]: no, it's `collate` just as the datapipe function name: `default_collate`

Comment (frozenbugs) [reply python/dgl/graphbolt/minibatch_sampler.py:14]: I see, but still typo.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 4 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/minibatch_sampler.py:129]: Worth to call it out in the comment.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 5 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/minibatch_sampler.py]: I did not see this comment addressed.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 5 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py]: it's addressed.

File: python/dgl/graphbolt/minibatch_sampler.py (hunk 4 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/minibatch_sampler.py:130]: comment is added.
